### DISCUSSION TITLE
Path sent to createFile by Windows is "\desktop.ini" when opening a fake file

### DISCUSSION THREAD

**Discussion by Thomas-Giaroli** [dokan-dev/dokany#1147] [Q&A]
Hello, i've been trying my hand on dokan-dotnet and i can't get rid of a problem i have with my implementation.

I've been trying to create a program with Dokan-dotnet that can accomplish these few steps :
1) Mount a disk on letter R
2) Create a fake file inside of it named "exemple.txt"
3) When the user tries to open the file "exemple.txt", create it.
4) Finally, open the newly created file.

I had no problems until i've reached step 3, i don't know why but it seems like i only get "\desktop.ini" sent to createFile instead of "exemple.txt", and this no matter what happens, it's the same thing when i open my mounted disk :R, and that's problematic cause i can't get my program to have the behavior i expect, since i never have the good filename.

You can find below my implementation of the FindFiles and ReadFile functions (FindFiles is used to create the "fake file" exemple.txt and readFile is where fileName is "\desktop.ini" instead of "\exemple.txt") : 

FindFiles:

        public NtStatus FindFiles(string fileName, out IList<FileInformation> files, IDokanFileInfo info)
        {
            files = new List<FileInformation>();
            var finfo = new FileInformation
            {
                FileName = "exemple.txt",
                Attributes = FileAttributes.Normal,
                LastAccessTime = DateTime.Now,
                LastWriteTime = null,
                CreationTime = DateTime.Now
            };
            files.Add(finfo);
            return DokanResult.Success;
        }
        
____________________________________________________________________________________________________________________________
        
ReadFile : 
        
        public NtStatus ReadFile(string fileName, byte[] buffer, out int bytesRead, long offset, IDokanFileInfo info)
        {
            var oh = "test\n";
            System.IO.File.AppendAllText(logFilePath, oh);
            System.IO.File.AppendAllText(logFilePath, "PATH = " + GetPath(fileName) + '\n');
            if (((System.IO.File.GetAttributes(fileName) & (FileAttributes.System | FileAttributes.Hidden)) != 0) || (fileName == "\\desktop.ini")) {
                System.IO.File.AppendAllText(logFilePath, "SYS FILE DETECTED\n");
                bytesRead = 0;
                return DokanResult.AccessDenied;
            }
            else {
            if (System.IO.File.Exists(GetPath(fileName)))
            {
                var message = "1er IF\n";
                System.IO.File.AppendAllText(logFilePath, message);
                if (info.Context == null) // memory mapped read
                {
                    using (var stream = new FileStream(GetPath(fileName), FileMode.Open, System.IO.FileAccess.Read))
                    {
                        stream.Position = offset;
                        bytesRead = stream.Read(buffer, 0, buffer.Length);
                    }
                }
                else // normal read
                {
                    var stream = info.Context as FileStream;
                    lock (stream) //Protect from overlapped read
                    {
                        stream.Position = offset;
                        bytesRead = stream.Read(buffer, 0, buffer.Length);
                    }
                }
            }
            else
            {
                var message = "2eme else\n";
                System.IO.File.AppendAllText(logFilePath, message);
                try
                {
                    using (FileStream fileStream = System.IO.File.Create(GetPath(fileName)))
                    {
                        fileStream.Close();
                        if (System.IO.File.Exists(GetPath(fileName)))
                        {
                            message = "THE FILE EXIST\n";
                            System.IO.File.AppendAllText(logFilePath, message);
                        }
                        else
                        {
                            message = "FILE DOES NOT EXIST\n";
                            System.IO.File.AppendAllText(logFilePath, message);
                        }
                    }
                }
                catch (Exception error)
                {
                    var ok = "CATCH, ERROR = " + error.Message + '\n';
                    System.IO.File.AppendAllText(logFilePath, ok);
                        bytesRead = 0;
                }
                using (var stream = new FileStream(GetPath(fileName), FileMode.Open, System.IO.FileAccess.Read))
                {
                    lock (stream) //Protect from overlapped read
                    {
                        stream.Position = offset;
                        bytesRead = stream.Read(buffer, 0, buffer.Length);
                    }
                }
            }
        }
            return (DokanResult.Success);
        }
        
____________________________________________________________________________________________________________________________

I'm already sorry if this not in any way related to Dokan-dotnet, I am currently learning C# and Dokan-dotnet simulteanously, and it's my first time ever developping anything on Windows.

Please let me know if any more information is needed, i really need to get this to work and i'll be more than grateful for any help.

Thanks already!

**Comment by Liryna** [comment]
Hi @Thomas-Giaroli ,

It is normal to get a Create for "\desktop.ini" request. Explorer is just doing it to see if there is one.
Are you capable of seeing "exemple.txt" when browsing with Explorer ?
I would suggest using [Procmon](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon) and look at all the requests to understand what is happening.

**Reply by Thomas-Giaroli** [reply]
Yes, i can see my file when browsing the R:\ disk, it's actually when i try to open it that i get desktop.ini instead of exemple.txt.
I am currently working on something else but i'll update this whenever i can launch it again and take a look at Procmon.

**Reply by Thomas-Giaroli** [reply]
So i've been trying my hand at this again, i used Procmon like you suggested and figured out that the paths when trying to access to example.txt were totally off, i get C:/example.txt instead of R:/example.txt.

By the way i only get one request looking for desktop.ini at path C:/desktop.ini before i get the multiple requests for example.txt, i don't know if that's a normal behaviour since i'm looking for things in R:/.

Could it be related to the fact that I'm using System.IO.file commands? Or it's maybe related to the fact that i don't seem to get the whole absolute path in the variable fileName when looking for something, and that may be why it's looking for everything i ask for at the root of the C:/ disk. (fileName for R:/example.txt is "\\example.txt" in debugger)

I'll attach below the different screenshots i took of Procmon, you'll maybe see something i didn't.

![1](https://github.com/dokan-dev/dokany/assets/72007344/51134825-3ba8-4611-a8e6-697c87e2cd07)

![2](https://github.com/dokan-dev/dokany/assets/72007344/82bfbbaf-046d-4289-a6b7-473038379435)

Thanks already

**Reply by Thomas-Giaroli** [reply]
By the way, i was wondering if it would be possible to access the FileInformation class for a file/directory but from the inside of the CreateFile function? I wanted to try something like setting the length of the file i create through FindFiles to a really specific negative value so i could use it as some kind of flag later.

To be honest i'm really struggling to figure out how to apply the behaviour i want, i think it is because of my lack of understanding of the Windows file system globally, so if you happen to have any kind of ressources regarding this too i'd gladly take it.

Sorry for being this needy but i can't find anything online to help me, and thanks for the support.

**Reply by Liryna** [reply]
Hey Tomas, from what I see `ConsoleApp1.exe` is your implementation of the filesystem. The CreateFile `C:\example.txt` is your implementation accessing this file which I imagine (if you are mirroring the data) is when you get a request from `R:\example.txt`.

With Procmon what I suggest is to look at the activity in your mount point. In your case is to add a filter for path starting with `R:\`.
